core/cli/src/informant/display.rs

```diff
@@ -0,0 +1,149 @@
+}
+
+impl<B: BlockT> InformantDisplay<B> {
+	/// Builds a new informant display system.
+	pub fn new() -> InformantDisplay<B> {
+		InformantDisplay {
+			last_number: None,
+			last_update: time::Instant::now(),
+		}
+	}
+
+	/// Displays the informant by calling `info!`.
+	pub fn display(&mut self, info: &ClientInfo<B>, net_status: NetworkStatus<B>) {
+		let best_number = info.chain.best_number;
+		let best_hash = info.chain.best_hash;
+		let speed = speed::<B>(best_number, self.last_number, self.last_update);
+		self.last_update = time::Instant::now();
+		self.last_number = Some(best_number);
+
+		let (status, target) = match (net_status.sync_state, net_status.best_seen_block) {
+			(SyncState::Idle, _) => ("Idle".into(), "".into()),
+			(SyncState::Downloading, None) => (format!("Syncing{}", speed), "".into()),
+			(SyncState::Downloading, Some(n)) => (format!("Syncing{}", speed), format!(", target=#{}", n)),
+		};
+
+		let finalized_number: u64 = info.chain.finalized_number.saturated_into::<u64>();
+
+		info!(
+			target: "substrate",
+			"{}{} ({} peers), best: #{} ({}), finalized #{} ({}), ⬇ {} ⬆ {}",
+			Colour::White.bold().paint(&status),
+			target,
+			Colour::White.bold().paint(format!("{}", net_status.num_connected_peers)),
+			Colour::White.paint(format!("{}", best_number)),
+			best_hash,
+			Colour::White.paint(format!("{}", finalized_number)),
+			info.chain.finalized_hash,
+			TransferRateFormat(net_status.average_download_per_sec),
+			TransferRateFormat(net_status.average_upload_per_sec),
+		);
+	}
+}
+
+/// Calculates `(best_number - last_number) / (now - last_update)` and returns a `String`
+/// representing the speed of import.
+fn speed<B: BlockT>(
+	best_number: NumberFor<B>,
+	last_number: Option<NumberFor<B>>,
+	last_update: time::Instant
+) -> String {
+	let elapsed = last_update.elapsed();
+	let since_last_millis = elapsed.as_secs() * 1000;
+	let since_last_subsec_millis = elapsed.subsec_millis() as u64;
+	let elapsed_ms = since_last_millis + since_last_subsec_millis;
+
+	// Number of blocks that have been imported since last time.
+	let diff = match last_number {
+		None => return String::new(),
+		Some(n) => best_number.saturating_sub(n)
+	};
+
+	if let Ok(diff) = TryInto::<u128>::try_into(diff) {
+		// If the number of blocks can be converted to a regular integer, then it's easy: just
+		// do the math and turn it into a `f64`.
+		let speed = diff.saturating_mul(10_000).checked_div(u128::from(elapsed_ms))
+			.map_or(0.0, |s| s as f64) / 10.0;
+		format!(" {:4.1} bps", speed)
+
+	} else {
+		// If the number of blocks can't be converted to a regular integer, then we need a more
+		// algebraic approach and we stay within the realm of integers.
+		let ten = (0..10)
+			.fold(<NumberFor<B> as Zero>::zero(), |a, _| a.saturating_add(One::one()));
+		let one_thousand = ten * ten * ten;
+		let elapsed = (0..elapsed_ms)
+			.fold(<NumberFor<B> as Zero>::zero(), |a, _| a.saturating_add(One::one()));
```

Comment:
I would expect that elapsed always fits into u32, so you should be able to also from.